fix: dapp guide (#196)
KILTprotocol/docs: docs/develop/07_dApp/02_well-known-did-config.md

---
id: well-known-did-config
title: Well Known DID Configuration
---

import TsJsSnippet from '@site/src/components/TsJsSnippet';
import TsJsBlock from '@site/src/components/TsJsBlock';

import DomainLinkageCtype from '!!raw-loader!@site/code_examples/dapp/src/dapp/domainLinkageCtype.ts';
import DomainLinkageClaim from '!!raw-loader!@site/code_examples/dapp/src/dapp/domainLinkageClaim.ts';
import SignPresentation from '!!raw-loader!@site/code_examples/dapp/src/dapp/signPresentation.ts';
import AttestCredential from '!!raw-loader!@site/code_examples/dapp/src/dapp/attestCredential.ts';
import FormatCredential from '!!raw-loader!@site/code_examples/dapp/src/dapp/formatCredential.ts';

:::danger This is a working draft

The KILT support of the *Well-Known DID Configuration* uses unpublished specifications and will change in the future.

:::

The *Well-Known DID Configuration* is implemented as a security measure when setting up the communication session between the dapp and extension.

> It ensures that the DID that the browser extension is communicating to is linked to the domain that is visited by the browser.

Reviewer: ```suggestion
It ensures that the DID the browser extension is communicating to is linked to the domain that is visited by the browser.
```

> This rule is currently enforced by the KILT DID Wallet reference implementation (Sporran Extension), but might be relaxed in the future.

Reviewer: ```suggestion
This rule is currently enforced by the KILT Wallet reference implementation (Sporran Extension), but might be relaxed in the future.
```

Is it a DID Wallet? Right now I think it's just a wallet.

The implementation is based on the [*Well-Known DID Configuration*](https://identity.foundation/.well-known/resources/did-configuration/) specified by the Decentralized Identity Foundation.

Once a communication session between a dapp and an extension is opened, the extension will query `<domain-name>/.well-known/did-configuration.json`.
This JSON-file must contain a credential presentation that conforms to the [Domain Linkage CType][CType-Domain-Linkage].

> ## Setup the Well-Known DID Configuration

Reviewer: ```suggestion
## Set up the Well-Known DID Configuration
```

For the *Well-Known DID Configuration* you need to go through the following steps:

0. Create a full DID
    * You will need the `assertionMethodKey` a.k.a. `attestationKey` for signing the credential
    * The `authenticationKey` is required for signing the transaction
1. Create a claim
2. Attest the claim
3. Create a presentation
4. Host the presentation on your website at `https://<your domain>/.well-known/did-configuration.json`

### Create a DID

Your dapp needs a DID to identify itself to the extension.

> If you don't have a DID yet, follow the [*create a full DID* guide](../01_sdk/02_cookbook/01_dids/02_full_did_creation.md).

Reviewer: ```suggestion
If your dapp does not have a DID yet, follow the [*create a full DID* guide](../01_sdk/02_cookbook/01_dids/02_full_did_creation.md).
```

> Make sure to create your DID with an `assertionMethodKey` so that you are able to issue attestations.

Reviewer: ```suggestion
Make sure to create the DID with an `assertionMethodKey` so that you are able to issue attestations.
```

### Making the claim

> After you got a DID, we can make a claim about that DID.

Reviewer: ```suggestion
After you get a DID, you can make a claim about that DID.
```

You? We? Your dapp? Pronouns should be made consistent.

The claim has to be based on the [Domain Linkage CType][CType-Domain-Linkage]:

<TsJsBlock>
  {DomainLinkageCtype}
</TsJsBlock>

The credential is built from the CType, claim contents, and your dapp's unique DID:

> <TsJsSnippet>

Reviewer: Did you mean to use the snippet? I think it is not ideal that every snippet now has a return statement. Either we hide the return line, or we show the whole function including its signature.

Author: yeah was also not super happy about that. But even though it's not pretty, it doesn't hurt readability. It shows what's the outcome of the snippet. So it should even help to understand the flow of the script better.

we can show the complete file if you want?

Reviewer: I wouldn't show the complete block since the `main` function takes some parameters which we don't necessarily want to show to the users. For me, the choice is either between the way things are now, or by removing the return line (which is possible). I prefer the latter, but I leave the final decision to you, as I don't have a strong opinion.

Author: [7907375](https://github.com/KILTprotocol/docs/pull/196/commits/7907375717ba778338261db1125faab29f31fb51)

  {DomainLinkageClaim}
</TsJsSnippet>

The credential isn't attested yet and is therefore not valid yet.

### Self-attesting the credential

A valid credential requires an attestation.

> Since we want to signal that the DID whats to link itself to the website, the DID has to attest to the claim itself.

Reviewer: ```suggestion
Since the website wants to link itself to the DID just created, it has to self-attest the domain linkage credential, i.e., write the credential attestation on chain using the same DID it is trying to link to.
```

The DID does not attest anything, since it's just an identifier. I tried to rephrase the sentence a bit to make it easier to read and also more technically correct.

In order to attest the credential we go through the following steps:

1. calculating the claim hash
2. creating the attest transaction
3. authorizing the transaction with your DID
4. paying for the transaction with a KILT account and submitting it to the chain

<TsJsSnippet>
  {AttestCredential}
</TsJsSnippet>

If you want to learn more about attestations you can refer to our [concept guide](../../concepts/04_credentials/04_attestation.md) or the [cookbook](../01_sdk/02_cookbook/04_claiming/03_attestation_creation.md)

### Presenting the credential

> Before we can host the credential on the website, we need to derive a presentation from it.

Reviewer: ```suggestion
To use the newly attested credential, we need to derive a presentation from it to host on the dapp website.
```

We host the presentation and not the credential, right?

<TsJsSnippet>
  {SignPresentation}
</TsJsSnippet>

> The Well Known DID Configuration specification requires a specific format (JSON-LD) for the domain linkage credential.

Reviewer: 1. Well Known, or Well-known?
2. Can we just say the credential needs to be a verifiable credential, linking to the spec? That implies it's a JSON-LD document.

Author: I think it's well-known what's correct 2f725b7c2f728cb053e43b6275f09671f5a64341

> For now we have to manually convert our credential into the required format.

Reviewer: ```suggestion
For now we have to manually convert our KILT credential into the required format.
```

<TsJsSnippet>
  {FormatCredential}
</TsJsSnippet>

### Host the Presentation

Now that you generated a presentation, you need to host it in your web app, so that the extension can query the presentation.
The extension will make an HTTP GET request to the following URI, and your dapp must respond with the presentation.

`/.well-known/did-configuration.json`

How the file is hosted depends on your project setup and is out of scope for this guide.

> [CType-Domain-Linkage]: https://github.com/KILTprotocol/ctype-index/tree/main/ctypes/0x9d271c790775ee831352291f01c5d04c7979713a5896dcf5e81708184cc5c643

Reviewer: Some links are collected here, some are not. Maybe we can collect them all here?

Author: 0d62d5fcbc8cc956a9f969ab5665eefb2d6daa11
